### PR TITLE
[Sumtree]: Prefix Sum Resync Condition

### PR DESCRIPTION
**Pull request by crnbarr93**
## What is the purpose of the change
In the sumtree mechanism there is a case where realizing a left node causes the right node to meet criteria to be realized. This is a tricky situation to cover without resyncing the node after a sync is complete (leading to O(n) worst case runtime). 

The naiive approach is to check if realizing the left node pushes the target ETAS over the right node's minimum boundary, but this causes a case where a left node may be synced multiple times (as the current implementation has no awareness whether or not a node has been realized). If a node is realized multiple times we incur a case where an order may be marked as filled when it is not and the price will run out of liquidity eventually as it is not in balance.

To counter the above we can give `prefix_sum_walk` context of what has been realized in the form of the previous sum. This opens the possibility of the following check:

1. Both left and right node exist
2. Check if anything of the left node has not been synced: 

```
Amount Realized Before Node = (Current Sum - Current Node Value)
Amount Realized Previously At Node = max(Previous Sum - Amount Realized Before Node, 0)
Unrealized From Left= Value of Left Child - Amount Realized Previously At Node
```

3. If syncing the left node pushes the target ETAS over the right minimum then we can walk right:

```
Right Min >= Left Realizable Amount + Target ETAS
```

This calculation should circumvent resyncing by realizing that resyncing will cause the right node to be included in the prefix sum. The implementation is as follows:

``` rust
if left_child.is_some() && right_child.is_some() {
        let left_child = left_child.clone().unwrap();
        let right_child = right_child.clone().unwrap();

        let sum_at_node = current_sum.checked_sub(node.get_value())?;
        let diff_at_node = prev_sum.saturating_sub(sum_at_node);
        let unrealized_from_left = left_child.get_value().saturating_sub(diff_at_node);
        let new_etas = target_etas.checked_add(unrealized_from_left)?;

        if new_etas >= right_child.get_min_range() {
            return prefix_sum_walk(storage, &right_child, current_sum, new_etas, prev_sum);
        }
}
```

## Testing and Verifying
Tests that used `prefix_sum_walk` were updated accordingly. Some have had the previous sum set as the current root node value (to simulate an explicit tree walk without the above condition) but should have test cases added in future for the above.

### CODE REVIEW

**Review by AlpinYukseloglu — COMMENTED** [review]
Doesn't this approach lead to the cancel after a partial fill being realized, meaning that the unfilled portion of that order can now be claimed?

I messaged offline what I believe would be a much simpler solution in line with our original mechanism without resyncing. If you agree with it, I'll create an issue we can proceed with that.

If my concern for this PR is correct, we should also investigate why the fuzz tests didn't pick the issue up.

File: contracts/sumtree-orderbook/src/sumtree/tree.rs [hunk 1]
```diff
@@ -88,13 +90,40 @@ fn prefix_sum_walk(
...
+    // To cover this we can determine how much of the left node has been realized, using this we can then determine
+    // if realizing what is unrealized from the left node will result in a new ETAS that is enough to realize the
+    // right node (to some extent)
+    if left_child.is_some() && right_child.is_some() {
+        let left_child = left_child.clone().unwrap();
+        let right_child = right_child.clone().unwrap();
+
+        // Calculate what the sum is before realizing the current node
+        let sum_at_node = current_sum.checked_sub(node.get_value())?;
+        // Calculate how much of the node has been realized in a previous sync
```
Comment (AlpinYukseloglu) [inline contracts/sumtree-orderbook/src/sumtree/tree.rs:111]:
```suggestion
        // Calculate the amount of cumulative realized cancellations *below the current node*  at the end of the
        // previous sync.
        //
        // Recall that `prev_sum` is the cumulative *global* amount that was realized at the end of the previous sync. 
        // 
        // Concretely, if this value is ever nonzero for a node, the amount corresponds exactly to the amount realized
        // below the node at the end of the previous sync. In all other cases, it will snap to zero due to the saturating sub.
```

File: contracts/sumtree-orderbook/src/sumtree/tree.rs [hunk 2]
```diff
@@ -88,13 +90,40 @@ fn prefix_sum_walk(
...
+        let diff_at_node = prev_sum.saturating_sub(sum_at_node);
+        // Calculate how much of the left node is unrealized
```
Comment (AlpinYukseloglu) [inline contracts/sumtree-orderbook/src/sumtree/tree.rs:113]:
```suggestion
        // `unrealized_from_left` is the amount in the left child that remained unrealized at the end of the previous
        // sync.
        //
        // If the left child is fully realized then the subtraction here will either be zero (if none of the right child is
        // realized) or negative (if some of the right child is realized), since `diff_at_node` = amount realized below left
        // and below right children. Saturating sub will ensure both of these cases snap to zero.
        // 
        // Thus, if this value is ever nonzero, it means that the left child had some unrealized cancels in it, and the
        // amount corresponds exactly to `unrealized_from_left`.
```

File: contracts/sumtree-orderbook/src/sumtree/tree.rs [hunk 3]
```diff
@@ -88,13 +90,40 @@ fn prefix_sum_walk(
...
+        let unrealized_from_left = left_child.get_value().saturating_sub(diff_at_node);
+        // Calculate the new ETAS after realizing what is unrealized from the left node
```
Comment (AlpinYukseloglu) [inline contracts/sumtree-orderbook/src/sumtree/tree.rs:115]:
```suggestion
        // Calculate the new ETAS after realizing what is unrealized from the left node
        //
        // Instead of doing this manually (and expensively) by triggering a resync, we simply add `unrealized_from_left`
        // to the running target ETAS value. This is functionally equivalent to simulating realizing the
        // remainder of the left child. As a sanity check, if it was already realized, then `unrealized_from_left`
        // would be zero and the target ETAS would remain unchanged.
```

File: contracts/sumtree-orderbook/src/sumtree/tree.rs [hunk 4]
```diff
@@ -88,13 +90,40 @@ fn prefix_sum_walk(
...
+        let new_etas = target_etas.checked_add(unrealized_from_left)?;
+
+        // if the new ETAS is greater than or equal to the right child's min range, we can walk right
+        // as the left node MUST be realizable given the invariants of the sumtree mechanism
```
Comment (AlpinYukseloglu) [inline contracts/sumtree-orderbook/src/sumtree/tree.rs:119]:
```suggestion
        // If the new ETAS is greater than or equal to the right child's min range, we can walk right
        // as the left node MUST be realizable given the invariants of the sumtree mechanism
        //
        // Intuition: If all swaps and cancels from the left child put us in the cancel territory of the right side, then we
        // don't care about the ordering of the swaps and cancels on the left. We know that all the cancels need to be
        // realized, so we batch realize them by adding *their unrealized portion* to our target ETAS and walking right.
```

**Review by AlpinYukseloglu — COMMENTED** [review]
After a long discussion offline, it seems your approach here is sound and my original intuition was mistaken. The two issues fixed here are:
1. The "batch realization" logic from the original sumtree mechanism was not included in the prefix sum walk
2. An optimization we made while implementing these contracts made our original design insufficient, and what you've designed & implemented seems to be the appropriate path forward.

I've made comment change suggestions to synthesize what this fix is actually doing, as well as a small request for higher clarity comments/reasoning for edge case tests. I suggest we do not modify existing unit tests at all to ensure they aren't broken by these changes.

Really great work with this!

File: contracts/sumtree-orderbook/src/tick.rs [hunk 5]
```diff
@@ -55,18 +55,26 @@ pub fn sync_tick(
...
+                    old_cumulative_realized_cancels, new_cumulative_realized_cancels
+                )),
+            })?;
 
         // Update the tick state to represent new ETAS and new cumulative realized cancels.
         tick_value.effective_total_amount_swapped = tick_value
             .effective_total_amount_swapped
             .checked_add(realized_since_last_sync)?;
         tick_value.cumulative_realized_cancels = new_cumulative_realized_cancels;
+        tick_value.last_tick_sync_etas = target_etas;
```
Comment (AlpinYukseloglu) [inline contracts/sumtree-orderbook/src/tick.rs:77]: Oh yikes nice catch here

**Review by AlpinYukseloglu — APPROVED** [review]
LGTM, great work with this.
